Permitted setting flags to action

MAUSMAUSgames opened the pull request:
I would love if I could use Action::_tag as a flag field.
It's a very small change to the code but a very powerful tool...

Use case:

```
    enum Flags {
        FRAME = 1,
        COLOR = 2,
        OPACITY = 4,
        POSITION = 8,
        SCALE = 16,
        ROTATION = 32,
        SKEW = 64
    };

    auto action1 = RepeatForever::create(Sequence::create(FadeTo::create( 0.1f, 155), 
                                                          TintTo::create(0.1f, 255,0,0),
                                                          nullptr));
    action1->setTag(Flags::OPACITY|Flags::COLOR);
    sprite->runAction(action1);

    auto action2 = RepeatForever::create(Sequence::create(Animate::create(idleAnimation),
                                                          MoveBy::create(0.3f, Vec2(5, 0)),
                                                          nullptr));
    action2->setTag(Flags::FRAME|Flags::POSITION);
    sprite->runAction(action2);

    // stop all actions that change position or opacity
    sprite->stopAllActionsByTag(Flags::OPACITY|Flags::POSITION, true); 
```

MAUSMAUSgames committed:
using Action::_tag as flag field

adding options to use tag as flag field to Node::stopAllActionsByTag
and ActionManager::removeAllActionsByTag

minggo commented:
What will happen if passing an unknown flag, such as 

sprite->stopAllActionsByTag(20, true);

MAUSMAUSgames commented:
Well, flags are single bits. So 20 would be 16 | 4. 
In my use case this would mean all Actions with Flags::OPACITY and Flags::SCALE would be stopped.

MAUSMAUSgames commented:
Just to be clear, it's up to the user how to use the flags. 

All I want is to add an option to `stopAllActionsByTag(int tag)` to stop all actions where `(Actions::_tag & tag) != 0` (bitwise AND)

zilongshanren commented:
@MAUSMAUSgames 

Sounds nice to me.

@ricardoquesada  Thoughts?

ricardoquesada commented:
the patch seems to be clean in the sense that by modifying only a few lines of code it as a new behavior to the code without breaking backwards compatibility.

But I prefer to add a new API instead of overriding `stopAllActionsByTag()`..

Instead I would add another one, like: `stopAllActionsByBitwiseTag()` or something like that to avoid confusion.

We also need some test cases in the patch.

zilongshanren commented:
Thanks, @ricardoquesada 
It makes sense and the suggested API is more cleaner.

Hi @MAUSMAUSgames  
Could you modify your PR according to @ricardoquesada  's comments?
Thanks.

MAUSMAUSgames committed:
added _flags to Action, change API to stopActionsByFlags

MAUSMAUSgames commented:
I introduced another member variable to Action: `unsigned int _flags` 
It defaults to to `0` (as opposed to `_tag` which is `int` and defaults to `-1 (Action::INVALID_TAG)`) which is more appropriate for a flag field

This is less invasive. Since I found that some classes set and use tags to actions internally  (`MenuItemLabel` and `ControlButton` use `kZoomActionTag`, and cocostudio seems to use tags on actions see `InnerActionFrame::onEnter()` in `CCFrame.cpp`).

I used the following APIs: 

```
void Action::setFlags(unsigned int flags);
unsigned int Action::getFlags();

void Node::stopActionsByFlags(unsigned int flags);
void ActionManager::removeActionsByFlags(unsigned int flags, Node *target);
```

Added a test case to `ActionManagerTest.cpp` but since I'm quite new to this, I'm not sure I did it the way it's supposed to be done.

milos1290 commented:
Great addition, exactly what i needed :)

zilongshanren commented on cocos/2d/CCActionManager.cpp:
```diff
@@ -315,6 +315,37 @@ void ActionManager::removeAllActionsByTag(int tag, Node *target)
+    {
+        return;
+    }
+    CCASSERT(target != nullptr, "");
```
We should say something in the CCASSERT macro, like "Target can't be nullptr".

ricardoquesada replied:
yeah.

MAUSMAUSgames replied:
I totally agree. 

Actually this is c&p-code from `ActionManager::removeAllActionsByTag()`. 
As a matter of fact, there are a several "empty" asserts of this kind in CCActionManager.cpp (same in CCActionInterval.cpp, CCActionEase.cpp, CCScheduler.cpp...)

Maybe someone should have a look at all of them and change them in one commit...

zilongshanren replied:
@MAUSMAUSgames  Thanks. I will change all the empty asserts.

zilongshanren replied:
@MAUSMAUSgames  Thanks. I will change all the empty asserts.
